GPT_SoVITS/AR/models/t2s_model.py

```diff
@@ -333,29 +338,64 @@ def infer_one_step(self, x, xy_attn_mask, k_cache, v_cache):
                 layer.self_attn.in_proj_bias
             ).chunk(3, dim=-1)
 
-            past_k = k_cache[layer_id]
-            past_v = v_cache[layer_id]
-
-            if past_k is not None:
-                k = torch.cat([past_k, k], 1)
-                v = torch.cat([past_v, v], 1)
-            k_cache[layer_id] = k
-            v_cache[layer_id] = v
-
             batch_size = q.shape[0]
             q_len = q.shape[1]
-            kv_len = k.shape[1]
-
-            q = q.view(batch_size, q_len, layer.self_attn.num_heads, -1).transpose(1, 2)
-            k = k.view(batch_size, kv_len, layer.self_attn.num_heads, -1).transpose(1, 2)
-            v = v.view(batch_size, kv_len, layer.self_attn.num_heads, -1).transpose(1, 2)
-
-            if xy_attn_mask is None:
-                attn = F.scaled_dot_product_attention(q, k, v)
+            
+            if flash_attn_with_kvcache is None:
+                past_k = k_cache[layer_id]
+                past_v = v_cache[layer_id]
+
+                if past_k is not None:
+                    k = torch.cat([past_k, k], 1)
+                    v = torch.cat([past_v, v], 1)
+                k_cache[layer_id] = k
+                v_cache[layer_id] = v
+                kv_len = k.shape[1]
+
+                q = q.view(batch_size, q_len, layer.self_attn.num_heads, -1).transpose(1, 2)
+                k = k.view(batch_size, kv_len, layer.self_attn.num_heads, -1).transpose(1, 2)
+                v = v.view(batch_size, kv_len, layer.self_attn.num_heads, -1).transpose(1, 2)
+
+                if xy_attn_mask is None:
+                    attn = F.scaled_dot_product_attention(q, k, v)
+                else:
+                    attn = F.scaled_dot_product_attention(q, k, v, ~xy_attn_mask)
+
+                attn = attn.permute(2, 0, 1, 3).reshape(-1, hidden_dim)
             else:
-                attn = F.scaled_dot_product_attention(q, k, v, ~xy_attn_mask)
+                q = q.view(batch_size, q_len, layer.self_attn.num_heads, -1)
+                k = k.view(batch_size, q_len, layer.self_attn.num_heads, -1)
+                v = v.view(batch_size, q_len, layer.self_attn.num_heads, -1)
+
+                if xy_attn_mask is None:
+                    attn = flash_attn_with_kvcache(q, k_cache[layer_id], v_cache[layer_id], k, v, cache_seqlens=cache_seqlens, causal=True)
+                else:
+                    # NOTE: there's a slight difference with the result produced by SDPA.
```

Comment:
Only the first inference has `xy_attn_mask`. This should not have a big impact on either inference speed or performance. However, if there's indeed a difference between the original SDPA and flash attention, in this case, I would vote for scrolling back to the original implementation since it won't impact the inference speed (only for the first inference).